fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/PartitionAvailability.java

```diff
@@ -0,0 +1,158 @@
+ * intersection of the availability of the child availabilities, taking into consideration only children whose rows
+ * are expected to participate.
+ * The default assumption is that the schemas of all tables will be the same, and if they are not, outer join
+ * semantics are generally expected. A partition function determines which availabilities will participate given a
+ * set of constraints, typically splitting on dimension values.
+ * <p>
+ * Note that this is for union across availabilities with the same dimension columns who's values are non-overlapping.
+ * For example, two availabilities of the following
+ * <pre>
+ * {@code
+ * +---------------------+----------------------+
+ * |      part 1         |       part 2         |
+ * +-------------------------+------------------+
+ * | [2017-01/2017-02]   |  [2017/2018]         |
+ * +---------------------+----------------------+
+ *
+ * If constraints participates in all parts:
+ *
+ * +-------------------------+
+ * |  availability(1,2)      |
+ * +-------------------------+
+ * | [2017-01/2017-02]       |
+ * +-------------------------+
+
+ * If a query participates only in the larger partition
+ *
+ * +-------------------------+
+ * |  availability(2)        |
+ * +-------------------------+
+ * | [2017-01/2018]          |
+ * +-------------------------+
+ * }
+ * </pre>
+ */
+public class PartitionAvailability extends BaseCompositeAvailability implements Availability {
+
+    private final Map<Availability, DataSourceFilter> availabilityFilters;
+    private final Set<TableName> dataSourceNames;
+    /**
+     * Constructor.
+     *
+     * @param availabilityFilters  Filters to identify which availabilities apply for a set of constraints
+     * Availabilities. The function is composed of the following works functions/transformations
+     * <pre>
+     * {@code
+     * 1. Function(DataSourceConstraint) -> Boolean
+     *
+     *
+     * 2. Function(ApiFilter) -> sef of PartitionKeys:
+     *        Dimension
+     *        Map<DimensionRow, PartitionKey>
+     *
+     *        SingleDimensionValueMap(Dimension, Map<DimensionRow, PartitionKey>)
+     *
+     *        apply():
+     *            dimension.getSearchProvider(ApiFilters)
+     *                    .findFilteredDimensionRowsPaged(
+     *                            apiFilters,
+     *                            PaginationParameters.EVERYTHING_IN_ONE_PAGE
+     *                    ).stream()
+     *                    .map(map::get)  // This will fail in potentially bad ways if the key isn't mapped
+     *                    .collect(Collectors.toSet());
+     *
+     * 3. Map<PartitionKey, Availability>
+     * }
+     * </pre>
+     */
+    public PartitionAvailability(
+            @NotNull Map<Availability, DataSourceFilter> availabilityFilters
+    ) {
+        this.availabilityFilters = availabilityFilters;
+        this.dataSourceNames = super.getDataSourceNames();
```

Comment:
Technically, it does not need to because of the dependency on getAllSourceAvailabilities(), but I'm happy to make this change since performance is a real factor on these classes.